File: services/query/src/identity.rs
//! This build's identity, recorded by the build script (roadmap R-121).

use crate::libs::wire::Timestamp;
use crate::proto::build::BuildInfo;

/// The component's directory name.
pub const COMPONENT: &str = "query";

/// The environment variable a deployment sets to the published SHA-256
/// of the running artifact. A binary cannot contain its own digest.
pub const ARTIFACT_SHA256: &str = "URSHANABI_ARTIFACT_SHA256";

/// The identity of the running build. Every value but the artifact
/// digest is fixed when the source is built; the digest is read from
/// the environment, and left empty unless it is a well-formed SHA-256.
///
/// # Panics
///
/// Never in practice: the build script refuses to build without a
/// positive integer commit time.
#[must_use]
pub fn this_build() -> BuildInfo {
    let committed: i64 = env!("URSHANABI_SOURCE_COMMITTED_AT")
        .parse()
        .expect("the build script records a positive integer");
    BuildInfo {
        component: COMPONENT.to_owned(),
        version: env!("URSHANABI_VERSION").to_owned(),
        source_revision: env!("URSHANABI_SOURCE_REVISION").to_owned(),
        source_committed_at: Some(Timestamp {
            seconds: committed,
            nanos: 0,
        }),
        artifact_sha256: artifact_sha256(std::env::var(ARTIFACT_SHA256).ok().as_deref()),
    }
}

/// Accepts a digest only if it is 64 lowercase hexadecimal characters.
fn artifact_sha256(value: Option<&str>) -> String {
    match value {
        Some(d) if d.len() == 64 && d.bytes().all(|b| matches!(b, b'0'..=b'9' | b'a'..=b'f')) => {
            d.to_owned()
        }
        _ => String::new(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_well_formed_digest_is_kept() {
        let d = "a".repeat(64);
        assert_eq!(artifact_sha256(Some(&d)), d);
    }

    #[test]
    fn a_missing_or_malformed_digest_is_empty() {
        assert_eq!(artifact_sha256(None), "");
        assert_eq!(artifact_sha256(Some("not-a-digest")), "");
        assert_eq!(artifact_sha256(Some(&"A".repeat(64))), "");
    }

    #[test]
    fn the_identity_describes_the_source() {
        let build = this_build();
        assert_eq!(build.component, "query");
        assert_eq!(build.source_revision.len(), 40);
        assert!(
            build
                .source_committed_at
                .is_some_and(|t| t.seconds > 0 && t.nanos == 0)
        );
    }
}

File: services/gateway/internal/identity/identity.go
// Package identity reports this build's identity (roadmap R-121). The
// source revision and commit time come from the build information the
// toolchain records in every binary built from a source repository, so
// two builds of one commit report one identity.
package identity

import (
	"errors"
	"os"
	"regexp"
	"runtime/debug"
	"time"
)

// Component is this component's directory name.
const Component = "gateway"

// ArtifactVariable names the environment variable a deployment sets to the
// published SHA-256 of the running artifact; a binary cannot contain its
// own digest.
const ArtifactVariable = "URSHANABI_ARTIFACT_SHA256"

// version is "development" unless a release build sets it with -ldflags -X.
var version = "development"

var sha256Hex = regexp.MustCompile(`^[0-9a-f]{64}$`)

// Build is one build's identity.
type Build struct {
	Component      string
	Version        string
	Revision       string
	CommittedAt    time.Time
	ArtifactSHA256 string
}

// Current returns the running binary's identity.
func Current() (Build, error) {
	info, ok := debug.ReadBuildInfo()
	if !ok {
		return Build{}, errors.New("the binary carries no build information")
	}
	return FromBuildInfo(info, os.Getenv(ArtifactVariable))
}

// FromBuildInfo reads an identity from recorded build information. The
// artifact digest is kept only if it is a well-formed SHA-256.
func FromBuildInfo(info *debug.BuildInfo, artifact string) (Build, error) {
	settings := map[string]string{}
	for _, s := range info.Settings {
		settings[s.Key] = s.Value
	}
	revision := settings["vcs.revision"]
	if len(revision) != 40 {
		return Build{}, errors.New("the binary was not built from a source repository")
	}
	committed, err := time.Parse(time.RFC3339, settings["vcs.time"])
	if err != nil {
		return Build{}, errors.New("the binary records no commit time")
	}
	if !sha256Hex.MatchString(artifact) {
		artifact = ""
	}
	return Build{
		Component:      Component,
		Version:        version,
		Revision:       revision,
		CommittedAt:    committed.UTC(),
		ArtifactSHA256: artifact,
	}, nil
}
